### PR TITLE
Move functions to base class

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** Some of these functions are needed in Catalyst. Since Catalyst does not use LM, but instead uses LightningDynamic, the least upper bound in the class hierarchy is `StateVectorLQubit`.

**Description of the Change:** Move these functions to `StateVetctorLQubit`.

**Benefits:** Code-reuse

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:711]: This was not in the original implementation, but we may want to reset the other state vector entries to `0.0`. Do we?

**Review by maliasadi — COMMENTED** [review]
@erick-xanadu What was the issue with adding these methods to `StateVectorLQubitDynamic`? Any plan to get this support for LK in a followup PR?

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (AmintorDusko) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:711]: I also wondered if we should check the indices size and compare it with the statevector size.

**Review by AmintorDusko — COMMENTED** [review]
The StateVectorLQubit has no explicit storage data type.
It might be tricky moving it here, or even unsafe.

**Comment by erick-xanadu** [comment]
> @erick-xanadu What was the issue with adding these methods to StateVectorLQubitDynamic? Any plan to get this support for LK in a followup PR?

The idea was to reuse code. So moving these to the base class is only for increasing code reuse. LK's implementation is already in the base class.

**Comment by erick-xanadu** [comment]
> The StateVectorLQubit has no explicit storage data type.
>  It might be tricky moving it here, or even unsafe.

Happy to close this ticket then. I'll leave it up for a bit of further discussion. The disadvantage is that we would copy these methods over. Right now for example, I am working on other two methods that will call these methods. And that means we would also have to copy these over in Catalyst.

**Comment by erick-xanadu** [comment]
@AmintorDusko alternatively, there could be another class between these two that implement these methods and Lightning Dynamic inherits from that one.

**Comment by vincentmr** [comment]
> The StateVectorLQubit has no explicit storage data type.

What is preventing us from getting a pointer to the data and modifying it? All the gates are applied in this way already, so isn't it natural to do the StatePrep similarly?

**Comment by AmintorDusko** [comment]
> > The StateVectorLQubit has no explicit storage data type.
> > It might be tricky moving it here, or even unsafe.
> 
> @AmintorDusko Happy to close this ticket then. I'll leave it up for a bit of further discussion. The disadvantage is that we would copy these methods over. Right now for example, I am working on other two methods that will call these methods. And that means we would also have to copy these over in Catalyst.

Hey @erick-xanadu, feel free to proceed and get the team's opinion on that.  
You don't need to close the PR. 
I agree that copying is an advantage, especially in the maintenance context.
We can improve safety by including some checks in the methods, like checking the indices size and comparing it with the statevector size, before proceeding. What do you think?

**Comment by AmintorDusko** [comment]
> > The StateVectorLQubit has no explicit storage data type.
> 
> What is preventing us from getting a pointer to the data and modifying it? All the gates are applied in this way already, so isn't it natural to do the StatePrep similarly?

Nothing. See my comment addressing Eric.

**Comment by erick-xanadu** [comment]
> We can improve safety by including some checks in the methods, like checking the indices size and comparing it with the statevector size, before proceeding. What do you think?

@AmintorDusko , added extra checks.

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:719]: Do we care that indices are unique (at this point of the pipeline)? If there are duplicates, the last occurence will have precedence.

Comment (erick-xanadu) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:719]: Your call :) Happy to remove this. I just added extra checks as it was asked for extra checks. Just let me know.

Comment (AmintorDusko) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:719]: I don't see a strong need for it either. Also, agree that an `unordered_set` for an array such as this will be heavy regarding performance and memory.

Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:719]: Let's remove it then.

File: .github/CHANGELOG.md
Comment (vincentmr) [inline .github/CHANGELOG.md:70]:
```suggestion
* Move `setBasisState`, `setStateVector` and `resetStateVector` from `StateVectorLQubitManaged` to `StateVectorLQubit`.
```

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:704]: Can we keep the previous implementation? 
```suggestion
        std::fill(arr, arr + length, 0.0);
```

**Review by vincentmr — COMMENTED** [review]
A couple suggestions, but I think it will be ready to merge after.

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (erick-xanadu) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:719]:
```suggestion
        auto num_indices = indices.size();
```

Comment (erick-xanadu) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp]: Removed here: https://github.com/PennyLaneAI/pennylane-lightning/pull/841/commits/c65256ff1680254547d1ed9e1dbce09e8f0f6ca7

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:720]: I think we should adopt L-Kokkos behaviour and [reset the SV to zero](https://github.com/PennyLaneAI/pennylane-lightning/blob/master/pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp#L155) before assigning new values.
```suggestion
std::fill(arr, arr + length, 0.0);
```

**Review by vincentmr — APPROVED** [review]
LGTM, thanks @erick-xanadu . Feel free to merge once the CI is green (I think you'll need to run `make format`).

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:723]: Do we need to normalize values after updating the indices or validate the new state-vector?

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:731]:
```suggestion
        if (this->getLength()) {
```

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:730]: Is this a noexcept method?
```suggestion
    void resetStateVector() noexcept {
```

**Review by maliasadi — COMMENTED** [review]
Thanks @erick-xanadu! :raised_hands:  Happy to approve after resolving my comment for `setStateVector`

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:723]: We could, but `qml.StatePrep` already validates the norm so I'd say we can skip it.

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:730]: It is calling `setBasisState` which can throw.

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (erick-xanadu) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:723]: > Do we need to normalize values after updating the indices or validate the new state-vector?

I am not familiar with the code base, so I don't know the answers to this question. I just moved the implementations and addressed the comments. If you find that the answer to this question is a yes, just let me know and we can normalize. I'll keep this conversation open and let you resolve it. :+1:

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (erick-xanadu) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:730]: Same.

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:723]: Sounds okay only if we ensure that this function is used from `qml.StatePrep`. I'll approve it now to unblock Erick and we can further discuss it offline :+1:

**Review by maliasadi — APPROVED** [review]

**Pull request merged**
